### ISSUE TITLE
Minor Project on Stock Prediction

### ISSUE THREAD

**Issue by Ashneet19** [Ashneet19/Corizo#1]
# Importing necessary libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn import metrics
from keras.models import Sequential
from keras.layers import Dense,LSTM,Dropout
# Importing data
df = pd.read_excel('1729258-1613615-Stock_Price_data_set_(1).xlsx')
df.head()

# Check shape of the dataset
df.shape

# Info of the dataset
df.info()

# Description of the dataset
df.describe()

# Sum of null values
df.isnull().sum()

# Looking for the unique values
df.nunique()

plt.figure(figsize=(15,5))
plt.plot(df['Close'], color="blue")
plt.title('Stock Close Price', fontsize=15)
plt.ylabel('Price in dollars')
plt.show()

# Splitting the data into training and testing sets

df_train = pd.DataFrame(df['Close'][0:int(len(df)*0.70)])             #70% used as a training data
df_test = pd.DataFrame(df['Close'][int(len(df)*0.70):int(len(df))])   #30% used as a testing data

print(df_train.shape)
print(df_test.shape)

# Checking the output of training & testing sets
df_train.head()

df_test.head()

# Scaling the data
scaler = MinMaxScaler(feature_range=(0,1))
df_train_array = scaler.fit_transform(df_train)
df_train_array

# Chekcking the shape of scaled array
df_train_array.shape

# Preparing the training data

X_train = []
y_train = []

for i in range(100,df_train_array.shape[0]):
    X_train.append(df_train_array[i-100:i])
    y_train.append(df_train_array[i,0])
    
X_train,y_train = np.array(X_train),np.array(y_train)
# Building model of 4 LSTM network followed by Dropout layout

model = Sequential()

model.add(LSTM(units=50, activation = 'relu', return_sequences = True, input_shape = (X_train.shape[1],1)))
model.add(Dropout(0.2))

model.add(LSTM(units=60, activation = 'relu', return_sequences = True))
model.add(Dropout(0.3))

model.add(LSTM(units=80, activation = 'relu', return_sequences = True))
model.add(Dropout(0.4))

model.add(LSTM(units=120, activation = 'relu'))
model.add(Dropout(0.5))

model.add(Dense(units = 1))
# Checking the summary
model.summary()

# Compiling & fitting the model 

model.compile(optimizer = 'adam', loss = 'mean_squared_error')
hist = model.fit(X_train,y_train, epochs = 50, batch_size = 32, verbose = 2 )

df_test.head()

df_train.tail()

# Append testing & training data
past_100_days = df_train.tail(100)
final_df = past_100_days.append(df_test, ignore_index=True)
# Scaling the data

input_data = scaler.fit_transform(final_df)
input_data

# Checking shape of the input_data
input_data.shape

# Preparing the testing data
X_test = []
y_test = []

for i in range(100,input_data.shape[0]):
    X_test.append(input_data[i-100:i])
    y_test.append(input_data[i,0])
    
X_test,y_test = np.array(X_test), np.array(y_test)
print(X_test.shape)
print(y_test.shape)

# Making Predictions

y_pred = model.predict(X_test)
print(y_pred.shape)

# Checking y_test
y_test

# Checking y_pred
y_pred

# Scaling the data
scaler.scale_

scale_factor = 1/0.00301214
y_pred = y_pred * scale_factor
y_test = y_test * scale_factor

# Plotting graph for the result
plt.figure(figsize = (15,5))
plt.plot(y_test,'b',label = 'Original Price')
plt.plot(y_pred,'r',label = 'Predicted Price')
plt.title('Stock Price Prediction')
plt.xlabel('Time')
plt.ylabel('Price')
plt.legend()
plt.show()
